internal/ioutil/ioutil.go

```diff
@@ -73,7 +73,54 @@ type ioret struct {
 	err error
 }
 
-// DeadlineWriter deadline writer with context
+// DeadlineReader deadline reader with timeout
+type DeadlineReader struct {
+	io.ReadCloser
+	timeout time.Duration
+	err     error
+}
+
+// NewDeadlineReader wraps a writer to make it respect given deadline
+// value per Write(). If there is a blocking write, the returned Reader
+// will return whenever the timer hits (the return values are n=0
+// and err=context.DeadlineExceeded.)
+func NewDeadlineReader(r io.ReadCloser, timeout time.Duration) io.ReadCloser {
+	return &DeadlineReader{ReadCloser: r, timeout: timeout}
+}
+
+func (r *DeadlineReader) Read(buf []byte) (int, error) {
+	if r.err != nil {
+		return 0, r.err
+	}
+
+	c := make(chan ioret, 1)
+	t := time.NewTimer(r.timeout)
+	go func() {
+		n, err := r.ReadCloser.Read(buf)
+		c <- ioret{n, err}
+		close(c)
+	}()
+
+	select {
+	case res := <-c:
+		if !t.Stop() {
+			<-t.C
+		}
+		r.err = res.err
+		return res.n, res.err
+	case <-t.C:
+		r.ReadCloser.Close()
```

Comment:
Since this is a very low level, I don't recommend closing this here, let's keep the one done in Close()

Golang *os.File.Close() returns 'file already closed' if it is called the second time.

Reply:
why? so that we can leak fd's? @vadmeste

Reply:
We should close here, why are you saying otherwise?

Reply:
> why? so that we can leak fd's? @vadmeste

umm normally it will not be leaked because NewDeadlineReader() returns io.ReadCloser and I think the idometic way to use it is similar to this:
```
f, err := os.Open(filename)
if err != nil {
    panic(err)
}
defer f.Close()
f.Read()...
```

but yeah I see now that we never check for the returned error from f.Close()